presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java

```diff
@@ -0,0 +1,285 @@
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+import static java.util.Objects.requireNonNull;
+
+public class AlluxioCachingFileSystem
+        extends ExtendedFileSystem
+{
+    private AlluxioCachingFileSystemInternal cachingFs;
+
+    public AlluxioCachingFileSystem(Configuration configuration, URI uri, ExtendedFileSystem fileSystem, CacheFactory cacheFactory)
+    {
+        this.cachingFs = new AlluxioCachingFileSystemInternal(configuration, uri, fileSystem, cacheFactory);
+    }
+
+    @Override
+    public synchronized void initialize(URI uri, Configuration conf)
+            throws IOException
+    {
+        requireNonNull(uri, "uri is null");
+        requireNonNull(conf, "conf is null");
+        cachingFs.initialize(uri, conf);
+    }
+
+    @Override
+    public String getScheme()
+    {
+        return cachingFs.getScheme();
+    }
+
+    @Override
+    public URI getUri()
+    {
+        return cachingFs.getUri();
+    }
+
+    @Override
+    public FSDataInputStream open(Path f, int bufferSize)
+            throws IOException
+    {
+        return cachingFs.open(f, bufferSize);
+    }
+
+    @Override
+    public FSDataOutputStream create(Path f, FsPermission permission, boolean overwrite, int bufferSize, short replication, long blockSize, Progressable progress)
+            throws IOException
+    {
+        return cachingFs.create(f, permission,
+                overwrite, bufferSize, replication, blockSize, progress);
+    }
+
+    @Override
+    public FSDataOutputStream append(Path f, int bufferSize, Progressable progress)
+            throws IOException
+    {
+        return cachingFs.append(f, bufferSize, progress);
+    }
+
+    @Override
+    public boolean rename(Path src, Path dst)
+            throws IOException
+    {
+        return cachingFs.rename(src, dst);
+    }
+
+    @Override
+    public boolean delete(Path f, boolean recursive)
+            throws IOException
+    {
+        return cachingFs.delete(f, recursive);
+    }
+
+    @Override
+    public FileStatus[] listStatus(Path f)
+            throws FileNotFoundException, IOException
+    {
+        return cachingFs.listStatus(f);
+    }
+
+    @Override
+    public void setWorkingDirectory(Path workingDirectory)
+    {
+        cachingFs.setWorkingDirectory(workingDirectory);
+    }
+
+    @Override
+    public Path getWorkingDirectory()
+    {
+        return cachingFs.getWorkingDirectory();
+    }
+
+    @Override
+    public boolean mkdirs(Path f, FsPermission permission)
+            throws IOException
+    {
+        return cachingFs.mkdirs(f, permission);
+    }
+
+    @Override
+    public FileStatus getFileStatus(Path f)
+            throws IOException
+    {
+        return cachingFs.getFileStatus(f);
+    }
+
+    @Override
+    public FSDataInputStream openFile(Path path, HiveFileContext hiveFileContext)
+            throws Exception
+    {
+        return cachingFs.open(path);
```

Comment:
If fileContext is not Cacheable then caching should not be used. hiveFileContext.isCacheable() would provide information at runtime if the caching should be used or not.

Reply:
is the switch between caching  & not caching happens at FileSystem level (instructed by `cacheConfig.isCachingEnabled()`) or changing at runtime even by the same FileSystem instance (indicated by `hiveFileContext.isCacheable()`)? My understanding was the former. If the latter one is correct,

Reply:
Both. Caching can be disabled at the config and run time level.